File: src/kalman.js
/**
* KalmanFilter
* @class
* @author Wouter Bulten
* @see {@link http://github.com/wouterbulten/kalmanjs}
* @version Version: 1.0.0-beta
* @copyright Copyright 2015-2018 Wouter Bulten
* @license MIT License
* @preserve
*/
export default class KalmanFilter {

  /**
  * Create 1-dimensional kalman filter
  * @param  {Number} options.R Process noise
  * @param  {Number} options.Q Measurement noise
  * @param  {Number} options.A State vector
  * @param  {Number} options.B Control vector
  * @param  {Number} options.C Measurement vector
  * @return {KalmanFilter}
  */
  constructor({R = 1, Q = 1, A = 1, B = 0, C = 1} = {}) {

    this.R = R; // noise power desirable
    this.Q = Q; // noise power estimated

    this.A = A;
    this.C = C;
    this.B = B;
    this.cov = NaN;
    this.x = NaN; // estimated signal without noise
  }

  /**
  * Filter a new value
  * @param  {Number} z Measurement. Either a value or an array with measurement and error.
  * @param  {Number} u Control
  * @return {Number}
  */
  filter(z, u = 0) {

    let zVal = z;
    let zErr = this.Q;

    if (Array.isArray(z)) {
      zVal = z[0];
      zErr = z[1];
    }

    if (isNaN(this.x)) {
      this.x = (1 / this.C) * zVal;
      this.cov = (1 / this.C) * zErr * (1 / this.C);
    }
    else {

      // Compute prediction
      const predX = this.predict(u);
      const predCov = this.uncertainty();

      // Kalman gain
      const K = predCov * this.C * (1 / ((this.C * predCov * this.C) + zErr));

      // Correction
      this.x = predX + K * (zVal - (this.C * predX));
      this.cov = predCov - (K * this.C * predCov);
    }

    return this.x;
  }

  /**
  * Predict next value
  * @param  {Number} [u] Control
  * @return {Number}
  */
  predict(u = 0) {
    return (this.A * this.x) + (this.B * u);
  }

  /**
  * Return uncertainty of filter
  * @return {Number}
  */
  uncertainty() {
    return ((this.A * this.cov) * this.A) + this.R;
  }

  /**
  * Return the last filtered measurement
  * @return {Number}
  */
  lastMeasurement() {
    return this.x;
  }

  /**
  * Set measurement noise Q
  * @param {Number} noise
  */
  setMeasurementNoise(noise) {
    this.Q = noise;
  }

  /**
  * Set the process noise R
  * @param {Number} noise
  */
  setProcessNoise(noise) {
    this.R = noise;
  }
}

File: contrib/java/kalman.py
#
# Simple Implementation of the Kalman Filter for 1D data, without any dependencies
# Originally written in JavaScript by Wouter Bulten
# Now rewritten in Python
#
# 2017
#
# license GNU LESSER GENERAL PUBLIC LICENSE v3
# author Sifan Ye
#
# see https://github.com/wouterbulten/kalmanjs
#

import math

class KalmanFilter:
    A = 1
    B = 0
    C = 1

    R = 0
    Q = 0

    cov = float('nan')
    x = float('nan')

    # Constructor
    #
    # R: Process Noise
    # Q: Measurement Noise
    def __init__(self, R, Q):
        self.R = R
        self.Q = Q

    # Filters a Measurement
    #
    # Measurement: The measurement value to be filtered
    # return: The filtered value
    def filter(self, measurement):
        u = 0
        if math.isnan(self.x):
            self.x = (1 / self.C) * measurement
            self.cov = (1 / self.C) * self.Q * (1 / self.C)
        else:
            predX = (self.A * self.x) + (self.B * u)
            predCov = ((self.A * self.cov) * self.A) + self.R

            # Kalman Gain
            K = predCov * self.C * (1 / ((self.C * predCov * self.C) + self.Q));

            # Correction
            self.x = predX + K * (measurement - (self.C * predX));
            self.cov = predCov - (K * self.C * predCov);

        return self.x

    # Returns the last measurement fed into the filter
    def last_measurement(self):
        return self.x

    # Sets measurement noise
    #
    # noise: The new measurement noise
    def set_measurement_noise(self, noise):
        self.Q = noise

    # Sets process noise
    #
    # noise: The new measurement noise
    def set_process_noise(self, noise):
        self.R = noise

# Main method as an example
def main():
    test = KalmanFilter(0.008, 0.1)
    testData = [66,64,63,63,63,66,65,67,58]
    for x in testData:
        print "Data:", x
        print "Filtered Data: ", test.filter(x)

main()